v2/callctx/callctx.go

```diff
@@ -74,9 +74,25 @@ func SetHeaders(ctx context.Context, keyvals ...string) context.Context {
 	h, ok := ctx.Value(headerKey).(map[string][]string)
 	if !ok {
 		h = make(map[string][]string)
+	} else {
+		h = cloneHeaders(h)
 	}
+
 	for i := 0; i < len(keyvals); i = i + 2 {
 		h[keyvals[i]] = append(h[keyvals[i]], keyvals[i+1])
 	}
 	return context.WithValue(ctx, headerKey, h)
 }
+
+// cloneHeaders makes a new key-value map while reusing the value slices.
+// As such, new values should be appended to the value slice, and modifying
+// indexed values is not thread safe.
+//
+// TODO: Replace this with maps.Clone when Go 1.21 is the minimum version.
+func cloneHeaders(h map[string][]string) map[string][]string {
+	c := make(map[string][]string, len(h))
+	for k, v := range h {
+		c[k] = v
```

Comment:
Is the issue the map itself, the slice, and or both? This only fixes the first case right?

Reply:
Right, I've described this in the comment. Since we only append to the value slice in SetHeaders, which returns a new slice I opted to not copy it.

Please check me on that though, also happy to do the slice/array copy.

Reply:
slice append only returns a new slice if the len would be > cap when adding the element. Or else it uses the same underlying array with the len attribute of the slice increased. But your test does currently still passes...

Reply:
Ah yes, that's right, I will change it copy to a new array/slice!

Reply:
Before you do try to find a test to make it race if you can. I tried for a couple of min and did not get one. But I think it should 🤷🏻